src/eval.rs

```diff
@@ -198,51 +204,70 @@ fn eval_expr(expr: &Expr, env: &Env, program: &Program) -> Result<Value, Runtime
             typ,
             overflow,
         } => {
-            let cast_t = match typ {
-                Type::BoolT => Err(RuntimeError::TypeError),
-                Type::SymT => Err(RuntimeError::TypeError),
-                Type::NumT(n) => Ok(n.clone()),
-            }?;
+            let Type::NumT(cast_t) = typ else {
+                return Err(RuntimeError::TypeError);
+            };
 
             let val = eval_expr(inner, env, &program)?;
 
             match val {
-                Value::Num(num, range) => match overflow {
-                    Overflow::Fail => cast(num, cast_t, Overflow::Fail),
-                    Overflow::Saturate => cast(num, cast_t, Overflow::Saturate),
-                    Overflow::Wraparound => cast(num, cast_t, Overflow::Wraparound),
-                },
+                Value::Num(num, _) => cast(num, cast_t.clone(), overflow.clone()),
                 // can only cast ints
-                Value::Bool(b) => Err(RuntimeError::TypeError),
-                Value::Sym(s) => Err(RuntimeError::TypeError),
+                _ => Err(RuntimeError::TypeError),
             }
         }
 
         Expr::Match { scrutinee, cases } => {
             let raw_val = eval_expr(&scrutinee, env, &program)?;
+            println!("raw val: {:?}", raw_val);
             for case in cases {
-                let pattern = match case.pattern {
-                    Pattern::Bool(b) => Value::Bool(b),
-                    Pattern::Num(n) => Value::Num(
-                        n,
-                        Range {
-                            start: n,
-                            end: n + 1,
-                        },
-                    ), // TODO: how would we match on Num, not including the range?
-                    Pattern::Sym(s) => Value::Sym(s),
-                    Pattern::Var(id) => env.get(&id).unwrap().clone(),
-                };
-                if raw_val == pattern {
-                    let g = eval_expr(&case.guard, env, &program).unwrap();
-
-                    if let Value::Bool(b) = g {
-                        if b {
-                            return Ok(eval_expr(&case.result, env, &program).unwrap());
+                println!("case {:?}", case);
+                match (case.pattern, raw_val.clone()) {
+                    (Pattern::Var(id), _) => {
+                        let mut env_scope = Env::new();
+                        env_scope.insert(id, raw_val.clone());
+                        let g = eval_expr(&case.guard, &env_scope, &program)?;
```

Comment:
there is a lot of duplication between this case and other cases. you should factor things out. e.g.:

```rust
let case_env = match (case.pattern, raw_val.clone() {
  // do environment adjustment and pattern matching
};
// eval the guard, if it is true eval the expression inside
// (basically all the repeated `let g = ...` code in all the cases)
```